src/Models/LagrangianParticleTracking/lagrangian_particle_advection.jl

```diff
@@ -48,27 +55,34 @@ bouncing the particle off the immersed boundary with a coefficient or `restituti
     X = flattened_node((x, y, z), ibg)
 
     # Determine current particle cell
-    fi, fj, fk = fractional_indices(X, ibg.underlying_grid, (c, c, c))
+    fi, fj, fk = fractional_indices(X, ibg.underlying_grid, c, c, c)
     i, j, k = truncate_fractional_indices(fi, fj, fk)
 
-    if immersed_cell(i, j, k, ibg)
-        # Determine whether particle was _previously_ in a non-immersed cell
-        i⁻, j⁻, k⁻ = previous_particle_indices
+    # Determine whether particle was _previously_ in a non-immersed cell
+    i⁻, j⁻, k⁻ = previous_particle_indices
 
-        if !immersed_cell(i⁻, j⁻, k⁻, ibg)
-            # Left-right bounds of the previous, non-immersed cell
-            xᴿ, yᴿ, zᴿ = node(i⁻+1, j⁻+1, k⁻+1, ibg, f, f, f)
-            xᴸ, yᴸ, zᴸ = node(i⁻,   j⁻,   k⁻,   ibg, f, f, f)
+    # Left bounds of the previous cell
+    xᴿ = xnode(i⁻ + 1, j, k, ibg, f, f, f)
+    yᴿ = ynode(i, j⁻ + 1, k, ibg, f, f, f)
+    zᴿ = znode(i, j, k⁻ + 1, ibg, f, f, f)
 
-            Cʳ = restitution
-            x⁺ = enforce_boundary_conditions(Bounded(), x, xᴸ, xᴿ, Cʳ)
-            y⁺ = enforce_boundary_conditions(Bounded(), y, yᴸ, yᴿ, Cʳ)
-            z⁺ = enforce_boundary_conditions(Bounded(), z, zᴸ, zᴿ, Cʳ)
+    # Right bounds of the previous cell
+    xᴸ = xnode(i⁻, j,  k,  ibg, f, f, f)
+    yᴸ = ynode(i,  j⁻, k,  ibg, f, f, f)
+    zᴸ = znode(i,  j,  k⁻, ibg, f, f, f)
```

Comment:
This is quite different from how it was written before. Was the previous implementation  bug?

Reply:
Yes, there was a bug with the previous implementation and it needed re-writing. Specifically, it was failing for `Flat` dimensions iirc. I'm not familiar with how this is implemented in curvilinear coords, and I don't remember lagrangian tracking being done in them before. Can we keep the code as it currently is in the PR?

Reply:
well, I understand why `node(i⁻+1, j⁻+1, k⁻+1, ibg, f, f, f)` would fail, but `node(i⁻, j, k, ibg, f, f, f)` should exist also for `Flat` topologies. Can we do this:
```julia
xᴸ, yᴸ, zᴸ = node(i⁻, j⁻, k⁻, ibg, f, f, f)
```
and keep the right boundaries as written in this PR?

Reply:
The previous code fails not because of the `+1` index in the flat dimension, but because it doesn't return anything for the flat dimension:

```
julia> grid_base = RectilinearGrid(topology = (Periodic, Flat, Periodic), size=(5, 5), extent=(1, 1))
5×1×5 RectilinearGrid{Float64, Periodic, Flat, Periodic} on CPU with 3×0×3 halo
├── Periodic x ∈ [0.0, 1.0)  regularly spaced with Δx=0.2
├── Flat y                   
└── Periodic z ∈ [-1.0, 0.0) regularly spaced with Δz=0.2

julia> node(2, 2, 2, grid_base, Center(), Center(), Center())
(0.3, -0.7)
```

So in these cases the statement `xᴸ, yᴸ, zᴸ = node(i⁻, j⁻, k⁻, ibg, f, f, f)` fails with

```
  Got exception outside of a @test
  BoundsError: attempt to access Tuple{Float64, Float64} at index [3]
  Stacktrace:
    [1] indexed_iterate
      @ ./tuple.jl:92 [inlined]
    [2] bounce_immersed_particle
      @ ~/repos/Oceananigans.jl2/src/Models/LagrangianParticleTracking/lagrangian_particle_advection.jl:75 [inlined]
```

Reply:
So, if I understand your comment correctly, we can just replace

```julia
    xᴿ = xnode(i⁻ + 1, j, k, ibg, f, f, f)
    yᴿ = ynode(i, j⁻ + 1, k, ibg, f, f, f)
    zᴿ = znode(i, j, k⁻ + 1, ibg, f, f, f)
```

with 

```julia
    xᴿ = xnode(i⁻ + 1, j⁻ + 1, k⁻ + 1, ibg, f, f, f)
    yᴿ = ynode(i⁻ + 1, j⁻ + 1, k⁻ + 1, ibg, f, f, f)
    zᴿ = znode(i⁻ + 1, j⁻ + 1, k⁻ + 1, ibg, f, f, f)
```

I think that should be equivalent. Note that `node()` is just a shorthand for `(xnode, ynode, znode)`

https://github.com/CliMA/Oceananigans.jl/blob/0f52a6a00db0cca06785df7ac5658638dd85d5a9/src/Grids/nodes_and_spacings.jl#L35-L37

Reply:
right, I am not completely sure this is needed for the `xᴿ`, `yᴿ` and `zᴿ` coordinates, but I think that the left ones need to be computed at the previous_particle_indices which is `i⁻, j⁻, k⁻`.

Reply:
Yeah I see that. Check the current version of the code. I just made a commit that should have the correct indices. Thanks for catching that!

Reply:
if you need either `x` or `λ` depending on the grid type you can use `ξnode` for example.

Reply:
I think it's better to use `ξnode` than `node[1]` which is a hack and not as easy to understand either.

Reply:
To clarify, I think there are at least 2 separate issues being discussed here.

1. The indices to be used when calling the location for the right indices. The code before this PR was `node(i⁻+1, j⁻+1, k⁻+1, ibg, f, f, f)`, which I believe should translate to `ξnode(i⁻+1, j⁻+1, k⁻+1, ibg, f, f, f)` and the same for `η` and `r`, with the same indices (that's how the PR actually is). @simone-silvestri seems to think that the correct call should be (as per a code suggestion below)

```julia
    xᴿ = xnode(i⁻ + 1, j⁻, k⁻, ibg, f, f, f)
    yᴿ = ynode(i⁻, j⁻ + 1, k⁻, ibg, f, f, f)
    zᴿ = znode(i⁻, j⁻, k⁻ + 1, ibg, f, f, f)
```
or perhaps with `ξ`, `η` and `r`, but importantly here is that the indices are different. @simone-silvestri, could you elaborate?

2. @glwagner is suggesting using `ξnode` (`η` and `r`) instead of `xnode` (and `y` and `z`). This is how it was in the code before this PR, but I think we actually want `xnode` and `ynode` since the advection is performed with velocities in `length/time`, no?

Reply:
I don't know what's correct. Seems like there is a missing test... but maybe @simone-silvestri can say. We did use this code on the sphere I thought.

Reply:
Ok, I am a bit confused which node should be the correct ``right'' node. I guess if there are tests now and the `i⁻+1, j⁻+1, k⁻+1, ibg, f, f, f` signature works, then the implementation is correct

Reply:
To be fair, tests works regardless of which of the two signatures we're considering here, so whatever the difference is, the tests don't pick up on it. I think I'll keep the algorithm as close to as it was before this PR as possible. This includes using the `i⁻+1, j⁻+1, k⁻+1, ibg, f, f, f` signature. As for the `xnode` versus `ξnode` debate, I actually realized the existing code used `xnode` for the non-immersed part of the algorithm and `ξnode` for the immersed grid. Likely the latter was done in error, so I'll keep everything as `xnode` for now. We can revisit this is the future if we find evidence that this isn't the proper way to do things.